src/transformers/generation_utils.py

```diff
@@ -39,6 +40,68 @@
 logger = logging.get_logger(__name__)
 
 
+@dataclass
+class GreedySearchDecoderOnlyOutput(ModelOutput):
+    """
+    Base class for outputs of encoder-decoder generation models using greedy search.
+
+    Args:
+        sequences (:obj: `torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`):
+            The generated sequences. The second dimension (sequence_length) is either equal to :obj:`max_length` or shorter if all
+            batches finished early due to the :obj:`eos_token_id`.
+        logits (:obj:`tuple(torch.FloatTensor)` `optional`, returned when ``output_scores=True`` is passed or when ``config.output_scores=True``):
+            Tuple of :obj:`torch.FloatTensor` of length :obj:`max_length` or shorter if all batches finished early, with each tensor of shape :obj:`(batch_size, config.vocab_size)`).
+        attentions (:obj:`tuple(tuple(torch.FloatTensor))`, `optional`, returned when ``output_attentions=True`` is passed or ``config.output_attentions=True``):
+            Tuple (one element for each generated token) of tuples (one element for each layer of the decoder) of :obj:`torch.FloatTensor` of shape
+            :obj:`(batch_size, num_heads, generated_length, sequence_length)`.
```

Comment:
There is a subtlety for generation that is a bit complex, but that should be documented here IMO. 

_Short context:_
For generation, usually only "decoder-only" models are used *a.k.a* GPT2, or the decoder of T5, Bart, etc...
"decoder-only" models use uni-directional self-attention meaning that previously projected past key and value states can be re-used and don't have to be calculated again => thus we can save computation this way. We have this feature enabled with the `use_cache` flag. So to compare what happens in the model, let's first see how generation works for `use_cache=False` (assuming that one word == one token):

step 1: [This is the prompt] -> decoder transfo -> [logit_1 logit_2 logit_3 logit_4] (each word was mapped to corresponding logit: logit_2 is output logit of "is" => in generation we sample next word from last logit (that's how LM works) => use logit_4 to sample <gen_word_1>
step 2: [This is the prompt <gen_word_1>] -> decoder transfo -> [logit_1(of this) logit_2 logit_3 logit_4 logit_5] => use logit_5 to sample <gen_word_2> 
step 3: [This is the prompt <gen_word_1> <get_word_2>] ...

Now however if `use_cache` is enabled the following happens:

step 1: [This is the prompt] -> decoder transfo -> [logit_1 logit_2 logit_3 logit_4] (each word was mapped to corresponding logit: logit_2 is output logit of "is" => in generation we sample next word from last logit (that's how LM works) => use logit_4 to sample <gen_word_1> **and** past key values of "This is the prompt" is returned as well. Here `past_key_values` has a length of 4
step 2: pass past_key_values **and**  [<gen_word_1>] -> decoder transfo -> [logit_5] => use logit_5 to sample <gen_word_2> **and** past key values are extended to have a length of 5
step 3: pass past_key_values of len 5 **and** [<gen_word_2>] ...
....

=> so now the relevant conclusion:

`hidden_states` has a length of `(batch_size, sequence_length, hidden_size)` if *no* `past_key_values` are passed, but has a length of `(batch_size, 1, hidden_size)` if past_key_values are passed. Similarly `attentions` is of size `(batch_size, num_heads, sequence_length, sequence_length)` and with `past_key_values` that are passed it's of size `(batch_size, num_heads, sequence_length, 1)`

Hope this helps a bit to understand the deal with `past_key_values`.

I'd suggest to slightly changed the description to 

` ... of shape :obj:`(batch_size, num_heads, sequence_length, sequence_length)` if no ``past_key_values`` are passed and of shape :obj:`(batch_size, num_heads, sequence_length, 1)` if ``past_key_values`` are passed. ``past_key_values`` are returned when ``use_cache=True``.

Reply:
Sorry for the long text :D But might be interesting to know how this feature works in the lib ;-)